### PR TITLE
Custom reason field - handling failure of auto-save when `back` button clicked

### PR DESCRIPTION
**Pull request by patmbolger**
PT Story: https://www.pivotaltracker.com/story/show/147708641

This code handles the situation where the user has:

1. Selected "Other" as the reason for waiting for info for a membership application (MA),
2. Is entering or editing a _custom-reason_, and,
3. Has not saved the _custom-reason_ (input focus is still in the text field - note that clicking _anywhere_ outside of the text field will cause a `change` event and thus save the _custom-reason_), and,
4) Clicks the _back_ button on the browser.

In this situation, the user will see an alert asking them if they want to leave the page (or similar message - will differ by browser - see below).

Other situations with the same setup (steps 1-3 above) behave as expected, such as:

- Clicking the _forward_ button on the browser causes the _custom-reason_ to be saved before going to the next page
- Clicking the "Cancel Waiting for Applicant Info" cancels that state and removes the _custom-reason_ from the DB
- Clicking any other link on the page causes the _custom-reason_ to be saved before going to the next view.

Note that - from what I have observed - it appears that the `change` event is being triggered in the above situations _before_ the current window is unloaded (except for cancel-waiting).  This causes unsaved _custom-reason_ text to be saved, and hence the callback described below does not prompt the user to confirm leaving the page.  (the _back_ button, however, does _not_ trigger a `change` event on the _custom-reason_ text field, thus callback detects unsaved changes and forces a user confirmation).

The key to this is the callback which is called before the window is unloaded (and before going to another page/window).  If that callback returns a string, the browser will show the confirmation popup for the user to confirm leaving the page or staying.

Evindently, on older browser versions, the string returned by the handler would be displayed to the user in the confirmation popup.  With newer versions, however, this is not the case:

"When this event returns (or sets the returnValue property to) a value other than null or undefined, the user is prompted to confirm the page unload. In some browsers, the return value of the event is displayed in this dialog. Starting with Firefox 4, Chrome 51, Opera 38 and Safari 9.1, a generic string not under the control of the webpage will be shown instead of the returned string." (https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload).

This is what is shown in these browsers:

### Chrome
chrome

---

### Firefox
firefox

---

### Safari
safari

---

I am assuming that the language used in the popup will conform to the user's browser settings or default locale (that is, our users would see Swedish here, not English).  @thesuss  - can you confirm?

I do not have access to a windows machine to test Internet Explorer.  Can someone do that?

If we go with this, I need to add tests of course.

Ready for review:
@weedySeaDragon @RobertCram @thesuss

### CODE REVIEW

File: app/views/membership_applications/_reason_waiting.html.haml [hunk 1]
```diff
@@ -28,13 +28,38 @@
...
-
 :javascript
 
-  $('#member_app_waiting_reasons').on('ajax:success', function (e, data) {
-    if (data === "#{@other_waiting_reason_value}") {
-      $('#other-text-field').show();
-    } else {
-      $('#other-text-field').hide();
-    }
+  var ChangeStatus = {
```
Comment (weedySeaDragon) [inline app/views/membership_applications/_reason_waiting.html.haml]: Any reason why this shouldn't be `let` ?

File: app/views/membership_applications/_reason_waiting.html.haml [hunk 2]
```diff
@@ -28,13 +28,38 @@
...
+    $('#custom_reason_text').on('change', function (e) {
+      ChangeStatus.unsaved_changes = false;
+    });
+
+    $(window).on('beforeunload', function (e) {
+
+      if ($('#member_app_waiting_reasons').val() === '#{@other_waiting_reason_value}' &&
+          ChangeStatus.unsaved_changes === true) {
+
+        e.returnValue = 'Unsaved changes! Do you want to leave the page?';
```
Comment (weedySeaDragon) [inline app/views/membership_applications/_reason_waiting.html.haml]: Will need to provide the i18n translation for this (on the chance that the browser is old and will actually show the message, of course).

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
Looks good.  Just the small note about translations and the minor code style comment.

File: app/views/membership_applications/_reason_waiting.html.haml (hunk 1 above)
Comment (patmbolger) [reply app/views/membership_applications/_reason_waiting.html.haml]: We're still on poltergeist for all javascript cucumber tests.  When we've expanded to include selenium/chrome as an option we can use the latest JS version.

File: app/views/membership_applications/_reason_waiting.html.haml (hunk 2 above)
Comment (patmbolger) [reply app/views/membership_applications/_reason_waiting.html.haml]: Good point - I will make that change and also add tests on the next revision of this PR.

**Comment by patmbolger** [comment]
I updated the PR with 1) cucumber test and 2) translation as suggested by @weedySeaDragon.

I am using some code from @zsuark's pending PR - I needed to use selenium/chrome to make the cucumber test work.

File: app/views/membership_applications/_reason_waiting.html.haml (hunk 2 above)
Comment (patmbolger) [reply app/views/membership_applications/_reason_waiting.html.haml]: Subsequent to this, I had to introduce Selenium/Chrome as a cucumber test option in order to get the added scenario (for this feature) to run.  So, I changed to `let` and also changed 3 existing scenarios to use selenium.

**Comment by RobertCram** [comment]
Looks good Patrick, if we want it to work this way. I myself would prefer a more "Apple like" approach, and have my changes saved unless I explicitly undo them. In this case that would mean just saving the changes on beforeunload. (That would also make forward- and backbutton behave the same) 

As it stands now (I'm working on Safari) I get a message telling me nothing about unsaved changes and not telling me why I would want to stay on the page. 

Not a big deal though. I'm ok with either.

**Review by RobertCram — APPROVED** [review]

**Comment by patmbolger** [comment]
@RobertCram - yes, I could go either way on this.  I chose this way because - in general - we have an "enter data and _explicitly_ save" motif pretty much everywhere else.

Given that, I thought that if the user had entered or deleted data, and _then_ hit the back button while still in the text entry field, then this is more likely a signal that the user wants to back out of the page without changes being saved (which is exactly what would happen as well if we had an explicit "save" button).

And yes, the Safari alert message is pretty useless.  FF and Chrome do a better job there.

File: Gemfile.lock [hunk 3]
```diff
@@ -120,11 +127,11 @@ GEM
       simplecov (<= 0.13)
     coderay (1.1.1)
     concurrent-ruby (1.0.5)
-    coveralls (0.8.21)
+    coveralls (0.8.19)
```
Comment (weedySeaDragon) [inline Gemfile.lock]: Some of your gems are older.  Are there particular reasons?  Gems where yours are older:
- `coveralls` 
- `simplecov`
- `thor`
(I think those are all of them)

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]

**Review by weedySeaDragon — APPROVED** [review]

**Pull request merged**